### PR TITLE
Universal refactor

### PR DESCRIPTION
**Pull request by dwalintukan**
Made the dataHex conversion more usable by everyone now. Can handle all the data types in Solidity, except for dynamic types (as recommended not to use by Qtum team).

### CODE REVIEW

File: dist/contract.js [hunk 1]
```diff
@@ -31,7 +31,11 @@ var SEND_GASLIMIT = 250000;
 var SEND_GASPRICE = 0.0000004;
 
 var MAX_BYTES_PER_ARRAY_SLOT = 64;
-var ARRAY_CAPACITY = 10;
+
+var REGEX_BYTES = /bytes([0-9]+)/;
```
Comment (frankobe) [inline dist/contract.js]: According to https://solidity.readthedocs.io/en/develop/types.html#reference-types,

Fixed-size byte arrays ranges from `bytes1` to `bytes32` only. So the regex can be stricter

File: dist/contract.js [hunk 2]
```diff
@@ -122,39 +126,43 @@ var Contract = function () {
 
       var hex = void 0;
       _lodash2.default.each(methodObj.inputs, function (item, index) {
```
Comment (frankobe) [inline dist/contract.js]: do we hv `_lodash1` or `_encoder1`, if not, can we rename it?

File: dist/contract.js [hunk 3]
```diff
@@ -122,39 +126,43 @@ var Contract = function () {
...
-              hex = _encoder2.default.addressToHex(args[index]);
+        var type = item.type;
+
+        if (type === 'address') {
+          hex = _encoder2.default.addressToHex(args[index]);
+          dataHex = dataHex.concat(hex);
+        } else if (type === 'bool') {
+          hex = _encoder2.default.boolToHex(args[index]);
+          dataHex = dataHex.concat(hex);
+        } else if (type.startsWith('uint')) {
```
Comment (frankobe) [inline dist/contract.js]: if we use regex (or type set as I suggestion) to match `bytes` type, probably I should match similarily on `int` & `uint` too

File: dist/contract.js [hunk 4]
```diff
@@ -122,39 +126,43 @@ var Contract = function () {
...
+              hex = _encoder2.default.stringToHex(args[index], MAX_BYTES_PER_ARRAY_SLOT * arrCapacity);
               dataHex = dataHex.concat(hex);
-              break;
             }
+          } else {
+            // fixed bytes, ie. bytes32
+            hex = _encoder2.default.stringToHex(args[index], MAX_BYTES_PER_ARRAY_SLOT);
+            dataHex = dataHex.concat(hex);
+          }
+        } else if (type === 'bytes') {
```
Comment (frankobe) [inline dist/contract.js]: Isnt `bytes` short for `bytes1`?

File: dist/encoder.js [hunk 5]
```diff
@@ -10,6 +10,10 @@ var _web3Utils = require('web3-utils');
 
 var _web3Utils2 = _interopRequireDefault(_web3Utils);
 
+var _bn = require('bn.js');
+
+var _bn2 = _interopRequireDefault(_bn);
```
Comment (frankobe) [inline dist/encoder.js]: The same question here, what's difference between `_bn` and `_bn2`

File: src/encoder.js [hunk 6]
```diff
@@ -57,6 +59,64 @@ class Encoder {
...
+   * @return The converted int to padded-left hex string.
+   */
+  static intToHex(num) {
+    if (_.isUndefined(num)) {
+      throw new Error(`num should not be undefined`);
+    }
+
+    // Must be converted to Two's Complement representation to handle negative numbers
+    const twosComp = new BN(num).toTwos(256).toJSON();
+    if (_.indexOf(num.toString()) === -1) {
```
Comment (frankobe) [inline src/encoder.js]: `if (num >= 0)`?

File: dist/contract.js (hunk 2 above)
Comment (chrisli30) [reply dist/contract.js]: These are auto-generated by babel. As far as I see, it's always _lodash2.

File: test/encoder.js [hunk 7]
```diff
@@ -1,129 +1,252 @@
...
+      "type": "function"
+    };
+
+    it('should convert a function obj to hash string', function() {
+      const hash = Encoder.getFunctionHash(funcObj);
       assert.equal(hash, '006a8a32');
       assert.equal(hash.length, 8);
     });
 
-    it('throws if methodObj is undefined', async function() {
```
Comment (chrisli30) [inline test/encoder.js]: Yeah, don't think "async" keyword is needed

File: src/encoder.js [hunk 8]
```diff
@@ -110,27 +170,13 @@ class Encoder {
...
-  }
-
   /*
    * Pads a hex string padded-left to 32 bytes.
    * @param {String} hexStr The hex string to pad.
    * @return {String} The padded-left hex string.
    */
   static padHexString(hexStr) {
-    if (hexStr === undefined) {
+    if (_.isUndefined(hexStr)) {
```
Comment (chrisli30) [inline src/encoder.js]: 👍

File: dist/contract.js (hunk 2 above)
Comment (dwalintukan) [reply dist/contract.js]: Yup auto-generated by babel when building.

File: dist/encoder.js (hunk 5 above)
Comment (dwalintukan) [reply dist/encoder.js]: Auto-generated by babel when building.

File: dist/contract.js (hunk 1 above)
Comment (dwalintukan) [reply dist/contract.js]: done. technically not necessary as the abi is generated from the contract and the contract won't compile if < 1 or > 32. but ya, someone can accidentally manually mess with the ABI types.

File: dist/contract.js (hunk 3 above)
Comment (dwalintukan) [reply dist/contract.js]: done

File: dist/contract.js (hunk 4 above)
Comment (dwalintukan) [reply dist/contract.js]: nope `bytes` is short for `bytes[]` dynamic type

File: src/encoder.js (hunk 6 above)
Comment (dwalintukan) [reply src/encoder.js]: fixed. good catch, the logic was wrong. I wasn't looking for the `-` sign. basically positive numbers get padded with `0` and negatives with `f` in Two's Complement.

File: test/encoder.js (hunk 7 above)
Comment (dwalintukan) [reply test/encoder.js]: :+1: erased all unnecessary async

**Pull request merged**